File: card-api/src/main/java/com/jean202/cardmizer/api/normalization/MerchantNormalizationRulesLoader.java
package com.jean202.cardmizer.api.normalization;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import org.yaml.snakeyaml.Yaml;

public class MerchantNormalizationRulesLoader {
    static final String DEFAULT_RESOURCE_PATH = "normalization/merchant-rules.yml";

    public MerchantNormalizationRules loadDefault() {
        return load(DEFAULT_RESOURCE_PATH);
    }

    public MerchantNormalizationRules load(String resourcePath) {
        Objects.requireNonNull(resourcePath, "resourcePath must not be null");

        try (InputStream inputStream = Thread.currentThread().getContextClassLoader().getResourceAsStream(resourcePath)) {
            if (inputStream == null) {
                throw new IllegalStateException("Normalization rules resource not found: " + resourcePath);
            }

            Object loaded = new Yaml().load(inputStream);
            if (!(loaded instanceof Map<?, ?> root)) {
                throw new IllegalStateException("Normalization rules must be a YAML object: " + resourcePath);
            }

            return new MerchantNormalizationRules(
                    stringMap(root.get("categoryAliases")),
                    stringMap(root.get("tagAliases")),
                    stringSet(root.get("simplePayTags")),
                    stringSet(root.get("onlineCategories")),
                    stringSet(root.get("offlineCategories")),
                    stringSet(root.get("subscriptionCategories")),
                    stringSet(root.get("digitalContentCategories")),
                    merchantRules(root.get("merchantRules"))
            );
        } catch (IOException exception) {
            throw new IllegalStateException("Failed to load normalization rules: " + resourcePath, exception);
        }
    }

    private static Map<String, String> stringMap(Object value) {
        if (value == null) {
            return Map.of();
        }
        if (!(value instanceof Map<?, ?> rawMap)) {
            throw new IllegalStateException("Expected YAML map but got: " + value.getClass().getSimpleName());
        }

        Map<String, String> result = new LinkedHashMap<>();
        for (Map.Entry<?, ?> entry : rawMap.entrySet()) {
            result.put(asString(entry.getKey()), asString(entry.getValue()));
        }
        return Map.copyOf(result);
    }

    private static Set<String> stringSet(Object value) {
        if (value == null) {
            return Set.of();
        }
        if (!(value instanceof List<?> rawList)) {
            throw new IllegalStateException("Expected YAML list but got: " + value.getClass().getSimpleName());
        }

        Set<String> result = new LinkedHashSet<>();
        for (Object item : rawList) {
            result.add(asString(item));
        }
        return Set.copyOf(result);
    }

    private static List<MerchantNormalizationRule> merchantRules(Object value) {
        if (value == null) {
            return List.of();
        }
        if (!(value instanceof List<?> rawList)) {
            throw new IllegalStateException("Expected YAML list but got: " + value.getClass().getSimpleName());
        }

        List<MerchantNormalizationRule> result = new ArrayList<>();
        for (Object item : rawList) {
            if (!(item instanceof Map<?, ?> rawRule)) {
                throw new IllegalStateException("Expected merchant rule object but got: " + item);
            }
            result.add(new MerchantNormalizationRule(
                    asString(rawRule.get("category")),
                    stringSet(rawRule.get("keywords")),
                    stringSet(rawRule.get("inferredTags"))
            ));
        }
        return List.copyOf(result);
    }

    private static String asString(Object value) {
        if (!(value instanceof String stringValue) || stringValue.isBlank()) {
            throw new IllegalStateException("Expected non-blank string but got: " + value);
        }
        return stringValue;
    }
}

File: card-api/src/main/kotlin/com/jean202/cardmizer/api/normalization/MerchantNormalizationRulesLoader.kt
package com.jean202.cardmizer.api.normalization

import org.yaml.snakeyaml.Yaml

class MerchantNormalizationRulesLoader {

    fun loadDefault(): MerchantNormalizationRules = load(DEFAULT_RESOURCE_PATH)

    fun load(resourcePath: String): MerchantNormalizationRules {
        val inputStream = Thread.currentThread().contextClassLoader.getResourceAsStream(resourcePath)
            ?: throw IllegalStateException("Normalization rules resource not found: $resourcePath")

        inputStream.use { stream ->
            val loaded = Yaml().load<Any>(stream)
            val root = loaded as? Map<*, *>
                ?: throw IllegalStateException("Normalization rules must be a YAML object: $resourcePath")

            return MerchantNormalizationRules(
                categoryAliases = stringMap(root["categoryAliases"]),
                tagAliases = stringMap(root["tagAliases"]),
                simplePayTags = stringSet(root["simplePayTags"]),
                onlineCategories = stringSet(root["onlineCategories"]),
                offlineCategories = stringSet(root["offlineCategories"]),
                subscriptionCategories = stringSet(root["subscriptionCategories"]),
                digitalContentCategories = stringSet(root["digitalContentCategories"]),
                merchantRules = merchantRules(root["merchantRules"]),
            )
        }
    }

    companion object {
        const val DEFAULT_RESOURCE_PATH = "normalization/merchant-rules.yml"

        private fun stringMap(value: Any?): Map<String, String> {
            value ?: return emptyMap()
            val rawMap = value as? Map<*, *>
                ?: throw IllegalStateException("Expected YAML map but got: ${value::class.simpleName}")
            return rawMap.entries.associate { asString(it.key) to asString(it.value) }
        }

        private fun stringSet(value: Any?): Set<String> {
            value ?: return emptySet()
            val rawList = value as? List<*>
                ?: throw IllegalStateException("Expected YAML list but got: ${value::class.simpleName}")
            return rawList.map { asString(it) }.toSet()
        }

        private fun merchantRules(value: Any?): List<MerchantNormalizationRule> {
            value ?: return emptyList()
            val rawList = value as? List<*>
                ?: throw IllegalStateException("Expected YAML list but got: ${value::class.simpleName}")
            return rawList.map { item ->
                val rawRule = item as? Map<*, *>
                    ?: throw IllegalStateException("Expected merchant rule object but got: $item")
                MerchantNormalizationRule(
                    category = asString(rawRule["category"]),
                    keywords = stringSet(rawRule["keywords"]),
                    inferredTags = stringSet(rawRule["inferredTags"]),
                )
            }
        }

        private fun asString(value: Any?): String {
            val s = value as? String
            require(!s.isNullOrBlank()) { "Expected non-blank string but got: $value" }
            return s
        }
    }
}
